git_machete/tests/functional/test_machete.py

```diff
@@ -26,6 +29,148 @@ def get_head_commit_hash() -> str:
+        self.headers: Dict[str, str] = headers
+        if method == "GET":
+            self.handle_get()
+        elif method == "PATCH":
+            self.handle_patch()
+        elif method == "POST":
+            self.handle_post()
+
+    def read(self) -> Union[str, bytes]:
+        return json.dumps(self.return_data).encode()
+
+    def handle_get(self) -> None:
+        if 'pulls' in self.url:
+            if '?head' in self.url:
+                head: str = self.url[self.url.rfind(':') + 1:]
+                for pr in GitAPIState.pulls:
+                    if pr['head']['ref'] == head:
+                        self.return_data = [pr]
+                        self.status_code = 200
+                        return
+                self.return_data = []
+                self.status_code = 404
+                return
+            else:
+                self.status_code = 200
+                self.return_data = GitAPIState.pulls
+                return
+        elif 'user' in self.url:
+            self.status_code = 200
+            self.return_data = GitAPIState.user
+            return
+
+    def handle_patch(self) -> None:
+        if 'issues' in self.url:
+            self.update_issue()
+        elif 'pulls' in self.url:
+            self.update_pull()
+
+    def handle_post(self) -> None:
+        if 'issues' in self.url:
+            if self.url.endswith('issues'):
+                self.update_issue(new=True)
+        elif 'pulls' in self.url:
+            if self.url.endswith('pulls'):
+                self.update_pull(new=True)
+
+    def update_pull(self, new: bool = False) -> None:
+        pull: Dict[str, Any]
+        if new:
+            pull = {'number': self.get_next_free_number(GitAPIState.pulls),
+                    'user': {'login': 'github_user'},
+                    'html_url': 'www.github.com'}
+        else:
+            pull_no: str = self.find_number(self.url)
+            pull = self.get_pull(pull_no)
+        for key in json.loads(self.json_data).keys():
+            if key in ('base', 'head'):
+                pull[key] = {'ref': ""}
+                pull[key]['ref'] = json.loads(self.json_data)[key]
+            else:
+                pull[key] = json.loads(self.json_data)[key]
+        GitAPIState.pulls.append(pull)
+        self.status_code = 201
+        self.return_data = pull
+
+    def update_issue(self, new: bool = False) -> None:
+        issue: Dict[str, Any]
+        if new:
+            issue = {'number': self.get_next_free_number(GitAPIState.issues)}
+        else:
+            issue_no = self.find_number(self.url)
+            issue = self.get_issue(issue_no)
+        for key in json.loads(self.json_data).keys():
+            issue[key] = json.loads(self.json_data)[key]
+        GitAPIState.issues.append(issue)
+        self.status_code = 201
+        self.return_data = issue
+        return
+
+    @staticmethod
```

Comment:
Hmmm there's quite a lot a static methods (and a lot of class attributes) in this class... this suggests that probably this class should be split into two classes, one with what's now static attributes but converted to instance attributes, and another with what's currently instance attributes. I'd imagine something like:

```python
class MockGitAPIState:
  # pulls, user, issues etc. are instance attributes
  def get_issue(...): ...
  # etc.
  def get_request() -> "MockGitAPIRequest":
   ...


class MockGitAPIRequest:
  def handle_get(...): ...
  # etc.
```

The question remains how to change `@mock.patch` annotations then... probably sth like:

```python
git_api_state_for_test_retarget_pr = [{'head': {'ref': 'feature'}, 'user': {'login': 'github_user'}, 'base': {'ref': 'root'}, 'number': '15', 'html_url': 'www.github.com'}]

@mock.patch('urllib.request.Request', git_api_state_for_test_retarget_pr.get_request)
....
```

**General idea**: the less global (incl. static) state, the better** (the less risk of interference b/w individual test cases)